python/cvmfs/repository.py

```diff
@@ -198,12 +200,170 @@ def _pop_catalog(self):
+
+
+    @staticmethod
+    def _write_content_into_file(content, opened_file):
+        """ Writes content into the opened file. The file must have write permission """
+        opened_file.write(content)
+        opened_file.seek(0)
+        opened_file.flush()
+
+
+    @abc.abstractmethod
+    def retrieve_file(self, file_name):
+        """ Abstract method to retrieve a file from the repository """
+        pass
+
+
+
+class LocalFetcher(Fetcher):
+    """ Retrieves files only from the local cache """
+
+    def __init__(self, cache_dir=''):
+        Fetcher.__init__(self, cache_dir, metadata_cleanup=False)
+
+    def retrieve_file(self, file_name):
+        """ The LocalFetcher works in read-only mode with its cache """
+        cached_file = self.cache.get(file_name)
+        if not cached_file:
+            raise FileNotFoundInRepository(self, file_name)
+        return cached_file
+
+
+
+class RemoteFetcher(Fetcher):
+    """ Retrieves files from the local cache if found, and from remote otherwise """
+
+    def __init__(self, repo_url, cache_dir=''):
+        Fetcher.__init__(self, cache_dir, metadata_cleanup=True)
+        self._repo_url = repo_url
+
+
+    def _download_content_and_store(self, cached_file, file_url):
+        response = requests.get(file_url, stream = True)
+        if response.status_code != requests.codes.ok:
+            raise FileNotFoundInRepository(self, file_url)
+        for chunk in response.iter_content(chunk_size=4096):
+            if chunk:
+                cached_file.write(chunk)
+        cached_file.seek(0)
+        cached_file.flush()
+        return cached_file
+
+
+    def __make_file_url(self, file_name):
+        return '/'.join([self._repo_url, file_name])
+
+
+    def _download_content_and_decompress(self, cached_file, file_url):
+        response = requests.get(file_url, stream=False)
+        if response.status_code != requests.codes.ok:
+            raise FileNotFoundInRepository(self, file_url)
+        decompressed_content = zlib.decompress(response.content)
+        Fetcher._write_content_into_file(decompressed_content, cached_file)
+        return cached_file
+
+
+    def _retrieve_file(self, file_name):
+        file_url = self.__make_file_url(file_name)
+        cached_file = self.cache.add(file_name)
+        return self._download_content_and_decompress(cached_file, file_url)
+
+
+    def retrieve_raw_file(self, file_name):
+        cached_file = self.cache.add(file_name)
+        file_url = self.__make_file_url(file_name)
+        return self._download_content_and_store(cached_file, file_url)
+
+
+    def retrieve_file(self, file_name):
+        cached_file = self.cache.get(file_name)
+        if cached_file:
+            return cached_file
+        return self._retrieve_file(file_name)
 
-    __metaclass__ = abc.ABCMeta
 
-    def __init__(self):
+
+class Repository:
+    """ Wrapper around a CVMFS Repository representation """
+
+    def __init__(self, cache_dir='', repo_url=''):
```

Comment:
Discussed offline. The parameters should be swapped and `repo_url` is supposed to be always required. However, it might contain different things:

* an HTTP URL (**http://cvmfs-stratum-zero.cern.ch/cvmfs/atlas.cern.ch**)
* a local path to a backend storage (**/srv/cvmfs/atlas.cern.ch**)
* a repository name (**atlas.cern.ch**) that needs a lookup in the config files